### PR TITLE
Conf files - v2

### PR DESCRIPTION
**Pull request by edersondisouza**
Changes from v1:
  - Accept `#` as a comment to Linuxfy a bit
  - New `strcasecmp` macro to compare strings, ignoring case
  - `null_check` macro is now `assert_or_return`.
  - New patch adding possibility to have a dir in which .conf files can be put, allowing one to add and override previous configuration. Files on this dir are read in alphabetical order - allowing one to chain overrides.
   -- This last patch involved some refactor - review it with care =D

### CODE REVIEW

**Comment by lucasdemarchi** [comment]
> -- This last patch involved some refactor - review it with care =D

where is the testsuite? :)

File: log.h [hunk 1]
```diff
@@ -61,7 +61,7 @@ int log_internal(int level, int error,
 #define log_warning_errno(error, ...)   log_full_errno(LOG_WARNING, error, __VA_ARGS__)
 #define log_error_errno(error, ...)     log_full_errno(LOG_ERR,     error, __VA_ARGS__)
 
-#define null_check(ptr, ...)                 \
+#define assert_or_return(ptr, ...)           \
```
Comment (lucasdemarchi) [inline log.h]: s/ptr/exp/ and handle it as an expression

**Comment by edersondisouza** [comment]
> where is the testsuite? :)

```
$ make test
/usr/bin/env python ./tests/test.py ./mavlink-routerd
ERROR LOADING MAVNATIVE - falling back to python implementation
All tests OK
```

=D

File: README.md [hunk 2]
```diff
@@ -43,18 +43,53 @@ Install:
 To route mavlink packets from master `ttyS1` to 2 other UDP endpoints, do as
 following:
 
-    $ mavlink-routerd -b 1500000 -e 192.168.7.1:14550 -e 127.0.0.1:14550 /dev/ttyS1
+    $ mavlink-routerd -e 192.168.7.1:14550 -e 127.0.0.1:14550 /dev/ttyS1:1500000
```
Comment (otaviobp) [inline README.md]: ++ for this change

File: conf.cpp [hunk 3]
```diff
@@ -121,7 +121,7 @@ int ConfFile::parse_file()
     return ret;
 }
 
-const char* ConfFile::next_section(bool reset)
+const char *ConfFile::next_section(bool reset)
```
Comment (otaviobp) [inline conf.cpp]: move to previous commit

File: conf.cpp [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    }
+
+    while ((read = getline(&entry, &n, file)) >= 0) {
+        line++;
+
+        _trim(&entry);
+
+        switch (entry[0]) {
+        case ';':
+        case '#':
```
Comment (lucasdemarchi) [inline conf.cpp]: people can just comment on the line above.   commenting inline in the end of the line just causes trouble... you would need to add a way to escape # when it's needed

Comment (edersondisouza) [inline conf.cpp]: Yeah - or to be more natural, accept quotes `"` to enclose a value, a need to escape them...
Not sure if worth to tackle now

File: conf.cpp [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+
+    char *end = strchr(entry, ']');
+    if (!end) {
+        log_error("On file %s: Line %d: Unfinished section name. Expected ']'", _filename, line);
+        return -EINVAL;
+    }
+
+    // So section name is the string between '[]'
+    *entry = ' ';
+    *end = '\0';
```
Comment (edersondisouza) [inline conf.cpp]: Do you mean like `[foo]bar`? Sounds fair

File: conf.cpp [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+            _current_section->current_config = config;
+            return config->value;
+        }
+        config = config->next;
+    }
+
+    return NULL;
+}
+
+void ConfFile::_trim(char **str)
```
Comment (edersondisouza) [inline conf.cpp]: Sure! when I started writing it I thought I would... hehe

**Comment by edersondisouza** [comment]
Updated to address @otaviobp comments.

File: conf.cpp [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        return -EINVAL;
+    }
+
+end:
+    free(entry);
+    fclose(file);
+    return ret;
+}
+
+const char* ConfFile::next_section(bool reset)
```
Comment (edersondisouza) [inline conf.cpp]: Hmmm... forgot this one. I'll try it.

**Comment by edersondisouza** [comment]
Now all comments should've been addressed.

File: conf.cpp (hunk 5 above)
Comment (otaviobp) [inline conf.cpp]: yep

File: main.cpp [hunk 8]
```diff
@@ -376,6 +463,68 @@ static int parse_conf()
...
+    if (ret < 0 && ret != -EIO) {
+        return ret;
+    }
+
+    dirname = get_conf_dir();
+    // Then, parse all files on configuration directory
+    dir = opendir(dirname);
+    if (!dir) {
+        return 0;
+    }
```
Comment (lucasdemarchi) [inline main.cpp]: Every return here one leaks dir.

**Comment by lucasdemarchi** [comment]
quickly scanned the changes, LGTM. There's certainly space for improvement, but it can be done later.

One small fix to do that I commented above. Please do on top.

File: main.cpp (hunk 8 above)
Comment (edersondisouza) [inline main.cpp]: ? This `return` happens when `dir` is `NULL`, maybe you commented on wrong line?

**Comment by lucasdemarchi** [comment]
I meant all returns below this line.

File: main.cpp (hunk 8 above)
Comment (lucasdemarchi) [inline main.cpp]: Here on... Like in "returns below this"

Comment (edersondisouza) [inline main.cpp]: Ah, maybe I'd have got 'from here' - sorry.
Will fix it - although any early return here leads to program termination.

**Pull request merged**